### PR TITLE
GH-36352: [Python] Add project_id to GcsFileSystem options

### LINKED ISSUE #36352: [Python][CI] GCFS python tests fail due to updated version of google-cloud-cpp
### Describe the bug, including details regarding any error messages, version, and platform.

As discussed here:
https://github.com/apache/arrow/pull/36119#issuecomment-1611147978

*originally from @raulcd *
I am unsure if there's something else to be done but it seems the nightlies for:
* [test-debian-11-python-3](https://github.com/ursacomputing/crossbow/runs/14609701083)
* [test-fedora-35-python-3](https://github.com/ursacomputing/crossbow/runs/14608852519)

Failed with:
```
E   pyarrow.lib.ArrowInvalid: google::cloud::Status(INVALID_ARGUMENT: missing project id error_info={reason=INVALID_ARGUMENT, domain=gcloud-cpp, metadata={gcloud-cpp.source.function=CreateBucket, gcloud-cpp.source.line=402, gcloud-cpp.source.filename=/build/cpp/google_cloud_cpp_ep-install/include/google/cloud/storage/client.h, gcloud-cpp.version=v2.12.0}}). Detail: [errno 22] Invalid argument
```
Is there something else to be done? Should I create a new issue for those failures?

and 
*originally from @raulcd *
No worries, from my understanding we have to add the new `project_id` to our `GcsFileSystem` constructor:
https://github.com/apache/arrow/blob/main/python/pyarrow/_gcsfs.pyx#L83-L89
And adding that to the underlying `CGcsOptions`.
We migt have to update the `__reduce__` function on that file and update the tests on `python/pyarrow/tests/test_fs.py`:
https://github.com/apache/arrow/blob/main/python/pyarrow/tests/test_fs.py#L211
I'll create an issue

### Component(s)

Continuous Integration, Python

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change
Some of our Python CI tests for GCS are failing due to the new project_id option added for GcsFileSystem here: https://github.com/apache/arrow/pull/36228

### What changes are included in this PR?

Added option

### Are these changes tested?

Will be tested on CI.

### Are there any user-facing changes?

Yes, there is a new project_id option when defining a GcsFileSystem.
* Closes: #36352

### CODE REVIEW

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

**Comment by raulcd** [comment]
@pitrou @jorisvandenbossche this is trying to fix a current failure on our nightly builds but I am unsure on how to exposed the `std::optional<std::string>` to Cython. I gave that a try but any pointer would be helpful as we can't seem to have other `optional[c_str]` around. See the current CI error here: https://github.com/apache/arrow/actions/runs/5411655158/jobs/9834688028?pr=36376#step:6:3752

File: python/pyarrow/_gcsfs.pyx [hunk 1]
```diff
@@ -136,6 +140,8 @@ cdef class GcsFileSystem(FileSystem):
         if retry_time_limit is not None:
             time_limit_seconds = retry_time_limit.total_seconds()
             options.retry_limit_seconds = time_limit_seconds
+        if project_id is not None:
+            options.project_id = tobytes(project_id)
```
Comment (jorisvandenbossche) [inline python/pyarrow/_gcsfs.pyx:144]:
```suggestion
            options.project_id = <c_string>tobytes(project_id)
```

Might help? (didn't test it)

Comment (raulcd) [reply python/pyarrow/_gcsfs.pyx]: Thanks that was good a pointer and solved the issue I was having!

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

**Review by pitrou — APPROVED** [review]
LGTM

**Comment by raulcd** [comment]
I don't recall seeing the `test-cuda-python` errors before but they don't seem related to this PR but to GDB tests

**Comment by raulcd** [comment]
I was expecting the `test-debian...`, `test-fedora-..` and `test-ubuntu-...` to succeed but there seems to be a new failure unrelated to this change:
```
________________ test_do_put_does_not_crash_when_schema_is_none ________________

    def test_do_put_does_not_crash_when_schema_is_none():
>       client = FlightClient('grpc+tls://localhost:9643',
                              disable_server_verification=True)

usr/local/lib/python3.9/dist-packages/pyarrow/tests/test_flight.py:2319: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyarrow/_flight.pyx:1275: in pyarrow._flight.FlightClient.__init__
    ???
pyarrow/_flight.pyx:1320: in pyarrow._flight.FlightClient.init
    ???
pyarrow/_flight.pyx:78: in pyarrow._flight.check_flight_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pyarrow.lib.ArrowNotImplementedError: Using encryption with server verification disabled is unsupported. Please use a release of Arrow Flight built with gRPC 1.27 or higher.
```
This PR is solving the original issue. I will open a new issue with the unrelated gRPC failure.

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
re-trigger as GH was down and almost all jobs failed.

**Comment by pitrou** [comment]
@raulcd Do you want to merge this?

**Comment by raulcd** [comment]
:+1:

**Pull request merged**